### PR TITLE
exercise 2

### PR DESCRIPTION
**Pull request by stevemao**

### CODE REVIEW

**Comment by stevemao** [comment]
I got warning:

```
$ stack ghc LogAnalysis
[2 of 2] Compiling LogAnalysis      ( LogAnalysis.hs, LogAnalysis.o )

LogAnalysis.hs:27:1: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
    In an equation for ‘insert’:
        Patterns not matched: (LogMessage _ _ _) (Node _ (Unknown _) _)
   |
27 | insert (Unknown _) tree = tree
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

File: cis194/week2/stevemao/LogAnalysis.hs [hunk 1]
```diff
@@ -22,3 +22,27 @@ parseMessage a = case words a of
...
+build :: [LogMessage] -> MessageTree
+build = foldr (\logMsg tree -> insert logMsg tree) Leaf
+
+inOrder :: MessageTree -> [LogMessage]
+inOrder Leaf = []
+inOrder (Node left a right) = inOrder left ++ [a] ++ inOrder right
+
+filterFn :: LogMessage -> Bool
+filterFn (LogMessage (Error severity) _ _) = severity > 50
+filterFn _ = False
```
Comment (shineli1984) [inline cis194/week2/stevemao/LogAnalysis.hs]: We can name this to isSevere

File: cis194/week2/stevemao/LogAnalysis.hs [hunk 2]
```diff
@@ -22,3 +22,27 @@ parseMessage a = case words a of
...
+
+insert :: LogMessage -> MessageTree -> MessageTree
+insert (Unknown _) tree = tree
+insert a Leaf = Node Leaf a Leaf
+insert log1@(LogMessage _ ts1 _) (Node left log2@(LogMessage _ ts2 _) right)
+  | ts1 > ts2 = Node left log2 (insert log1 right)
+  | otherwise = Node (insert log1 left) log2 right
+
+build :: [LogMessage] -> MessageTree
+build = foldr (\logMsg tree -> insert logMsg tree) Leaf
```
Comment (shineli1984) [inline cis194/week2/stevemao/LogAnalysis.hs]: `\a b -> f a b` is the same as `f`

File: cis194/week2/stevemao/LogAnalysis.hs [hunk 3]
```diff
@@ -22,3 +22,27 @@ parseMessage a = case words a of
...
+inOrder :: MessageTree -> [LogMessage]
+inOrder Leaf = []
+inOrder (Node left a right) = inOrder left ++ [a] ++ inOrder right
+
+filterFn :: LogMessage -> Bool
+filterFn (LogMessage (Error severity) _ _) = severity > 50
+filterFn _ = False
+
+mapper :: LogMessage -> String
+mapper (LogMessage _ _ message) = message
```
Comment (shineli1984) [inline cis194/week2/stevemao/LogAnalysis.hs]: We can name this getMessage.
Bit more importantly this function is partial!

Comment (stevemao) [reply cis194/week2/stevemao/LogAnalysis.hs]: It shouldn't be possible to match `Unknown`. Let me have a look how this should be done.

**Comment by stevemao** [comment]
If `foldr` is better than recursion I can revert. Thanks!

**Pull request merged**